jdbc-v2/src/test/java/com/clickhouse/jdbc/StatementTest.java

```diff
@@ -471,4 +477,28 @@ public void testConnectionExhaustion() throws Exception {
             }
         }
     }
+
+    @Test(groups = { "integration" })
+    public void testConcurrentCancel() throws Exception {
+        try (Connection conn = getJdbcConnection()) {
+            try (StatementImpl stmt = (StatementImpl) conn.createStatement()) {
+                stmt.executeQuery("SELECT number FROM system.numbers LIMIT 10000000000");
+                stmt.cancel();
+            }
+            try (StatementImpl stmt = (StatementImpl) conn.createStatement()) {
+                new Thread(() -> {
+                    try {
+                        ResultSet rs = stmt.executeQuery("SELECT number FROM system.numbers LIMIT 10000000000");
+                        rs.next();
+                        log.info(rs.getObject(1).toString());
+                    } catch (SQLException e) {
+                        log.error("Error in thread", e);
+                    }
+                }).start();
+
+                Thread.sleep(1000);
```

Comment:
no need to sleep - just repeat this 100 times in the loop:
1. start query 
2. cancel